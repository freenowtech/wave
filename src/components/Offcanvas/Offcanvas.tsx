import React, { useEffect, useState, ReactNode, useContext } from 'react';
import { createGlobalStyle } from 'styled-components';
import { WidthProps } from 'styled-system';
import { useIsEscKeyPressed } from '../../utils/hooks/useIsEscKeyPressed';
import { ANIMATION_DURATION as CARD_ANIMATION_DURATION, CenteredCard } from './components/CenteredCard';
import { ANIMATION_DURATION as CARD_ANIMATION_DURATION_SIDE, SideCard } from './components/SideCard';
import { ANIMATION_DURATION as DIMMING_ANIMATION_DURATION, DimmingFade } from './components/DimmingFade';
import { TopRightXIcon } from './components/TopRightXIcon';

type DismissFunc = () => void;

const DismissContext = React.createContext<DismissFunc>(undefined);

const useOffcanvasDismiss = (): DismissFunc => {
    const dismiss = useContext(DismissContext);

    if (dismiss === undefined) {
        throw new Error('useOffcanvasDismiss must be used within a <Offcanvas />');
    }

    return dismiss;
};

interface OffcanvasProps extends WidthProps {
    /**
     * Show the Offcanvas on the side of the page: left or right
     */
    side?: 'left' | 'right' | 'top' | 'bottom';
    /**
     * Makes the Offcanvas dismissible by the user (defaults to true)
     */
    dismissible?: boolean;
    /**
     * Function that will be called when the user intends to close the Offcanvas
     */
    onClose?: () => void;
    /**
     * Children can be rendered with a dismiss function as render prop. Calling
     * the dismiss function will close the Offcanvas.
     */
    children?: ReactNode | ((dismiss: DismissFunc) => ReactNode);
}

const PreventBackgroundScroll = createGlobalStyle`
    body {
        overflow: hidden;
    }
`;

const ANIMATION_DURATION = Math.max(
    DIMMING_ANIMATION_DURATION,
    CARD_ANIMATION_DURATION || CARD_ANIMATION_DURATION_SIDE
);

/*
 * We define the types for the props twice because the render props won't be inferred correctly
 * when only using `React.FC<OffcanvasProps>`. This leads to compiler errors when passing the
 * dismiss function.
 */
const Offcanvas: React.FC<OffcanvasProps> = ({
    children,
    onClose,
    dismissible,
    side = 'left',
    ...rest
}: OffcanvasProps) => {
    const [visible, setVisible] = useState(true);
    const isEscKeyPressed = useIsEscKeyPressed();

    const handleClose: DismissFunc = () => {
        setVisible(false);

        if (onClose) {
            setTimeout(() => onClose(), ANIMATION_DURATION);
        }
    };

    const handleDimmingClick = () => {
        if (dismissible) handleClose();
    };

    useEffect(() => {
        if (dismissible && isEscKeyPressed) {
            handleClose();
        }
    }, [dismissible, isEscKeyPressed]);

    const renderChildren = () => {
        if (typeof children === 'function') {
            return children(handleClose);
        }

        return children;
    };

    return (
        <DismissContext.Provider value={handleClose}>
            <DimmingFade onClick={handleDimmingClick} visible={visible} data-testid="dimming-background" />
            {side === 'top' || side === 'bottom' ? (
                <CenteredCard visible={visible} {...rest} side={side}>
                    {dismissible && <TopRightXIcon data-testid="close-icon" onClick={handleClose} />}
                    {renderChildren()}
                </CenteredCard>
            ) : (
                <SideCard visible={visible} {...rest} side={side}>
                    {dismissible && <TopRightXIcon data-testid="close-icon" onClick={handleClose} />}
                    {renderChildren()}
                </SideCard>
            )}
            <PreventBackgroundScroll />
        </DismissContext.Provider>
    );
};

Offcanvas.defaultProps = {
    dismissible: true
};

export { Offcanvas, OffcanvasProps, useOffcanvasDismiss };
