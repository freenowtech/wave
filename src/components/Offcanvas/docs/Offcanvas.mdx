---
name: Offcanvas
menu: Components
route: /components/Offcanvas
---

import { Offcanvas } from '../Offcanvas.tsx';
import { OffcanvasCreator } from './OffcanvasCreator.tsx';
import { OffcanvasPropsTable } from './OffcanvasPropsTable.tsx';
import { StyledSystemLinks } from '../../../docs/StyledSystemLinks'

# Offcanvas

The Offcanvas should only contain content or flows which are related to the current page and not replace pages, therefore
its complexity level has to be simple. Consider the fullscreen state if you want to focus the users attention to the Offcanvas
content.

Offcanvas is responsive per default but the width can be adjusted (the default is 600px or 37.5rem). Content on the Offcanvas
is scrollable and has a space5 bounding box.

<OffcanvasCreator />

### Example Code

The following contains a simple example, with a resulting Offcanvas similar to the live examples above.

```jsx
const [showOffcanvas, setShowOffcanvas] = useState(true);

<Offcanvas onClose={() => setShowOffcanvas(false)} fullscreen={true}>
    {dismiss => (
        <>
            <Headline as="h2">Add Note</Headline>
            <Text as="p" my={3}>
                Ensure that all participants are aware of the change.
            </Text>
            <Button onClick={dismiss}>Add Note</Button>
            <TextButton onClick={dismiss}>Cancel</TextButton>
        </>
    )}
</Offcanvas>;
```

## API

<OffcanvasPropsTable />

<br />

<StyledSystemLinks component="Offcanvas" supportedProps={ ['width'] }/>

### Dismiss

Offcanvas provides a function to dismiss the component with an animation. The
dismiss function is available either through render props or a hook. After the
animation has finished the `onClose` callback will be called.

#### Render Props

```jsx
<Offcanvas>
    {dismiss => <Button onClick={dismiss}>Close</Button>}
</Offcanvas>
```

#### Hook

```jsx
const InsideOffcanvas = () => {
    const dismiss = useOffcanvasDismiss();
    return <button onClick={dismiss}>Close</button>;
}

const OffcanvasWrapper = () => (
    <Offcanvas>
        <InsideOffcanvas />
    </Offcanvas>
);
```
