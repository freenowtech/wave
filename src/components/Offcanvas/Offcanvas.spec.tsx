import { fireEvent, render, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as React from 'react';
import { Offcanvas, useOffcanvasDismiss } from './Offcanvas';

describe('Offcanvas', () => {
    it('renders the default props', () => {
        expect(render(<Offcanvas>Content</Offcanvas>).container).toMatchSnapshot();
    });

    it('renders the left side variation', () => {
        expect(render(<Offcanvas side="left">Content</Offcanvas>).container).toMatchSnapshot();
    });

    it('renders the right side variation', () => {
        expect(render(<Offcanvas side="right">Content</Offcanvas>).container).toMatchSnapshot();
    });

    it('renders the top side variation', () => {
        expect(render(<Offcanvas side="top">Content</Offcanvas>).container).toMatchSnapshot();
    });

    it('renders the bottom side variation', () => {
        expect(render(<Offcanvas side="bottom">Content</Offcanvas>).container).toMatchSnapshot();
    });

    describe('should call onClose function', () => {
        it('when clicking on close icon', async () => {
            const mockCloseHandler = jest.fn();
            const { getByTestId } = render(<Offcanvas onClose={mockCloseHandler}>Content</Offcanvas>);

            fireEvent.click(getByTestId('close-icon'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });

        it('when clicking on the dimming background', async () => {
            const mockCloseHandler = jest.fn();
            const { getByTestId } = render(<Offcanvas onClose={mockCloseHandler}>Content</Offcanvas>);

            fireEvent.click(getByTestId('dimming-background'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });

        it('when pressing the escape key', async () => {
            const mockCloseHandler = jest.fn();
            render(<Offcanvas onClose={mockCloseHandler}>Content</Offcanvas>);

            fireEvent.keyUp(window, { keyCode: 27 });

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });

        it('when calling the dismiss function', async () => {
            const mockCloseHandler = jest.fn();
            const { getByText } = render(
                <Offcanvas onClose={mockCloseHandler}>
                    {dismiss => (
                        <button type="button" onClick={dismiss}>
                            dismiss
                        </button>
                    )}
                </Offcanvas>
            );

            fireEvent.click(getByText('dismiss'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });

        it('when calling the dismiss function from a hook', async () => {
            const mockCloseHandler = jest.fn();

            const InnerComponent: React.FC = () => {
                const dismiss = useOffcanvasDismiss();
                return (
                    <button type="button" onClick={dismiss}>
                        Click Me
                    </button>
                );
            };

            render(
                <Offcanvas onClose={mockCloseHandler}>
                    <InnerComponent />
                </Offcanvas>
            );

            userEvent.click(screen.getByRole('button'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });
    });

    describe('in the non-dismissible state', () => {
        it('should not show the close icon for the default Offcanvas', () => {
            const { queryByTestId } = render(<Offcanvas dismissible={false} />);

            expect(queryByTestId('close-icon')).toBeNull();
        });

        it('should not call the onClose handler when clicking on the dimming', async () => {
            const mockCloseHandler = jest.fn();
            const { getByTestId } = render(<Offcanvas dismissible={false} onClose={mockCloseHandler} />);

            fireEvent.click(getByTestId('dimming-background'));

            await waitFor(() => expect(mockCloseHandler).not.toHaveBeenCalled());
        });

        it('should not call the onClose handler when pressing the escape key', async () => {
            const mockCloseHandler = jest.fn();
            render(
                <Offcanvas dismissible={false} onClose={mockCloseHandler}>
                    Content
                </Offcanvas>
            );

            fireEvent.keyUp(window, { keyCode: 27 });

            await waitFor(() => expect(mockCloseHandler).not.toHaveBeenCalled());
        });

        it('should call the onClose handler when calling the dismiss function', async () => {
            const mockCloseHandler = jest.fn();
            const { getByText } = render(
                <Offcanvas onClose={mockCloseHandler} dismissible={false}>
                    {dismiss => (
                        <button type="button" onClick={dismiss}>
                            dismiss
                        </button>
                    )}
                </Offcanvas>
            );

            fireEvent.click(getByText('dismiss'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });
    });
});
