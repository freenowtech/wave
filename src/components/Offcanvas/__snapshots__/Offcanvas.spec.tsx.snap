// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offcanvas renders the bottom side variation 1`] = `
.c1 {
  overflow: auto;
  box-sizing: border-box;
  background-color: #FFFFFF;
  border-radius: 0.125rem;
  padding: 2.5rem;
  height: 28.375rem;
  box-shadow: 0 0 0.75rem 0.25rem rgba(0,0,0,0.12);
  position: fixed;
  left: 0;
  right: unset;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
  z-index: 1100;
  width: 100%;
  will-change: transform,opacity;
  -webkit-transition: -webkit-transform 150ms ease-out,opacity 150ms ease-out;
  -webkit-transition: transform 150ms ease-out,opacity 150ms ease-out;
  transition: transform 150ms ease-out,opacity 150ms ease-out;
  left: 0;
  right: 0;
  top: unset;
  bottom: 0;
}

.c1.centered-card-animation-appear,
.c1.centered-card-animation-enter {
  opacity: 0;
  -webkit-transform: translate(0%,calc(50% + 1rem));
  -ms-transform: translate(0%,calc(50% + 1rem));
  transform: translate(0%,calc(50% + 1rem));
}

.c1.centered-card-animation-appear-active,
.c1.centered-card-animation-enter-active {
  opacity: 1;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
}

.c1.centered-card-animation-exit {
  opacity: 1;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
}

.c1.centered-card-animation-exit-done,
.c1.centered-card-animation-exit-active {
  opacity: 0;
  -webkit-transform: translate(0%,10%);
  -ms-transform: translate(0%,10%);
  transform: translate(0%,10%);
}

.c0 {
  background-color: rgba(0,15,31,0.6);
  height: 100%;
  left: 0;
  opacity: 0.6;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1000;
  will-change: opacity;
  -webkit-transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
  transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
}

.c0.dimming-fade-animation-appear,
.c0.dimming-fade-animation-enter {
  opacity: 0;
}

.c0.dimming-fade-animation-appear-active,
.c0.dimming-fade-animation-enter-active {
  opacity: 0.6;
}

.c0.dimming-fade-animation-exit {
  opacity: 0.6;
}

.c0.dimming-fade-animation-exit-done,
.c0.dimming-fade-animation-exit-active {
  opacity: 0;
}

.c2 {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
  cursor: pointer;
  z-index: 1100;
}

@media (max-width:calc( + 2rem)) {
  .c1 {
    width: calc(100% - 2rem);
  }
}

<div>
  <div
    class="sc-AxhCb c0 dimming-fade-animation-appear dimming-fade-animation-appear-active"
    data-testid="dimming-background"
  />
  <div
    class="sc-AxjAm c1 centered-card-animation-appear centered-card-animation-appear-active"
    height="28.375rem"
  >
    <svg
      class="c2"
      color="#001E3E"
      data-testid="close-icon"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        fill="currentColor"
        fill-rule="nonzero"
      />
    </svg>
    Content
  </div>
</div>
`;

exports[`Offcanvas renders the default props 1`] = `
.c1 {
  overflow: auto;
  box-sizing: border-box;
  background-color: #FFFFFF;
  border-radius: 0.125rem;
  padding: 2.5rem;
  width: 28.375rem;
  box-shadow: 0 0 0.75rem 0.25rem rgba(0,0,0,0.12);
  position: fixed;
  left: 0;
  right: unset;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
  z-index: 1100;
  height: 100%;
  will-change: transform,opacity;
  -webkit-transition: -webkit-transform 150ms ease-out,opacity 150ms ease-out;
  -webkit-transition: transform 150ms ease-out,opacity 150ms ease-out;
  transition: transform 150ms ease-out,opacity 150ms ease-out;
  left: 0;
  right: unset;
  top: 0;
  bottom: 0;
}

.c1.centered-card-animation-appear,
.c1.centered-card-animation-enter {
  opacity: 0;
  -webkit-transform: translate(calc(-50% - 1rem),0%);
  -ms-transform: translate(calc(-50% - 1rem),0%);
  transform: translate(calc(-50% - 1rem),0%);
}

.c1.centered-card-animation-appear-active,
.c1.centered-card-animation-enter-active {
  opacity: 1;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
}

.c1.centered-card-animation-exit {
  opacity: 1;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
}

.c1.centered-card-animation-exit-done,
.c1.centered-card-animation-exit-active {
  opacity: 0;
  -webkit-transform: translate(-10%,0%);
  -ms-transform: translate(-10%,0%);
  transform: translate(-10%,0%);
}

.c0 {
  background-color: rgba(0,15,31,0.6);
  height: 100%;
  left: 0;
  opacity: 0.6;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1000;
  will-change: opacity;
  -webkit-transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
  transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
}

.c0.dimming-fade-animation-appear,
.c0.dimming-fade-animation-enter {
  opacity: 0;
}

.c0.dimming-fade-animation-appear-active,
.c0.dimming-fade-animation-enter-active {
  opacity: 0.6;
}

.c0.dimming-fade-animation-exit {
  opacity: 0.6;
}

.c0.dimming-fade-animation-exit-done,
.c0.dimming-fade-animation-exit-active {
  opacity: 0;
}

.c2 {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
  cursor: pointer;
  z-index: 1100;
}

@media (max-width:calc(28.375rem + 2rem)) {
  .c1 {
    width: calc(100% - 2rem);
  }
}

<div>
  <div
    class="sc-AxhCb c0 dimming-fade-animation-appear dimming-fade-animation-appear-active"
    data-testid="dimming-background"
  />
  <div
    class="sc-AxjAm c1 centered-card-animation-appear centered-card-animation-appear-active"
    width="28.375rem"
  >
    <svg
      class="c2"
      color="#001E3E"
      data-testid="close-icon"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        fill="currentColor"
        fill-rule="nonzero"
      />
    </svg>
    Content
  </div>
</div>
`;

exports[`Offcanvas renders the left side variation 1`] = `
.c1 {
  overflow: auto;
  box-sizing: border-box;
  background-color: #FFFFFF;
  border-radius: 0.125rem;
  padding: 2.5rem;
  width: 28.375rem;
  box-shadow: 0 0 0.75rem 0.25rem rgba(0,0,0,0.12);
  position: fixed;
  left: 0;
  right: unset;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
  z-index: 1100;
  height: 100%;
  will-change: transform,opacity;
  -webkit-transition: -webkit-transform 150ms ease-out,opacity 150ms ease-out;
  -webkit-transition: transform 150ms ease-out,opacity 150ms ease-out;
  transition: transform 150ms ease-out,opacity 150ms ease-out;
  left: 0;
  right: unset;
  top: 0;
  bottom: 0;
}

.c1.centered-card-animation-appear,
.c1.centered-card-animation-enter {
  opacity: 0;
  -webkit-transform: translate(calc(-50% - 1rem),0%);
  -ms-transform: translate(calc(-50% - 1rem),0%);
  transform: translate(calc(-50% - 1rem),0%);
}

.c1.centered-card-animation-appear-active,
.c1.centered-card-animation-enter-active {
  opacity: 1;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
}

.c1.centered-card-animation-exit {
  opacity: 1;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
}

.c1.centered-card-animation-exit-done,
.c1.centered-card-animation-exit-active {
  opacity: 0;
  -webkit-transform: translate(-10%,0%);
  -ms-transform: translate(-10%,0%);
  transform: translate(-10%,0%);
}

.c0 {
  background-color: rgba(0,15,31,0.6);
  height: 100%;
  left: 0;
  opacity: 0.6;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1000;
  will-change: opacity;
  -webkit-transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
  transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
}

.c0.dimming-fade-animation-appear,
.c0.dimming-fade-animation-enter {
  opacity: 0;
}

.c0.dimming-fade-animation-appear-active,
.c0.dimming-fade-animation-enter-active {
  opacity: 0.6;
}

.c0.dimming-fade-animation-exit {
  opacity: 0.6;
}

.c0.dimming-fade-animation-exit-done,
.c0.dimming-fade-animation-exit-active {
  opacity: 0;
}

.c2 {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
  cursor: pointer;
  z-index: 1100;
}

@media (max-width:calc(28.375rem + 2rem)) {
  .c1 {
    width: calc(100% - 2rem);
  }
}

<div>
  <div
    class="sc-AxhCb c0 dimming-fade-animation-appear dimming-fade-animation-appear-active"
    data-testid="dimming-background"
  />
  <div
    class="sc-AxjAm c1 centered-card-animation-appear centered-card-animation-appear-active"
    width="28.375rem"
  >
    <svg
      class="c2"
      color="#001E3E"
      data-testid="close-icon"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        fill="currentColor"
        fill-rule="nonzero"
      />
    </svg>
    Content
  </div>
</div>
`;

exports[`Offcanvas renders the right side variation 1`] = `
.c1 {
  overflow: auto;
  box-sizing: border-box;
  background-color: #FFFFFF;
  border-radius: 0.125rem;
  padding: 2.5rem;
  width: 28.375rem;
  box-shadow: 0 0 0.75rem 0.25rem rgba(0,0,0,0.12);
  position: fixed;
  left: unset;
  right: 0;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
  z-index: 1100;
  height: 100%;
  will-change: transform,opacity;
  -webkit-transition: -webkit-transform 150ms ease-out,opacity 150ms ease-out;
  -webkit-transition: transform 150ms ease-out,opacity 150ms ease-out;
  transition: transform 150ms ease-out,opacity 150ms ease-out;
  left: unset;
  right: 0;
  top: 0;
  bottom: 0;
}

.c1.centered-card-animation-appear,
.c1.centered-card-animation-enter {
  opacity: 0;
  -webkit-transform: translate(calc(50% + 1rem),0%);
  -ms-transform: translate(calc(50% + 1rem),0%);
  transform: translate(calc(50% + 1rem),0%);
}

.c1.centered-card-animation-appear-active,
.c1.centered-card-animation-enter-active {
  opacity: 1;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
}

.c1.centered-card-animation-exit {
  opacity: 1;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
}

.c1.centered-card-animation-exit-done,
.c1.centered-card-animation-exit-active {
  opacity: 0;
  -webkit-transform: translate(+10%,0%);
  -ms-transform: translate(+10%,0%);
  transform: translate(+10%,0%);
}

.c0 {
  background-color: rgba(0,15,31,0.6);
  height: 100%;
  left: 0;
  opacity: 0.6;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1000;
  will-change: opacity;
  -webkit-transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
  transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
}

.c0.dimming-fade-animation-appear,
.c0.dimming-fade-animation-enter {
  opacity: 0;
}

.c0.dimming-fade-animation-appear-active,
.c0.dimming-fade-animation-enter-active {
  opacity: 0.6;
}

.c0.dimming-fade-animation-exit {
  opacity: 0.6;
}

.c0.dimming-fade-animation-exit-done,
.c0.dimming-fade-animation-exit-active {
  opacity: 0;
}

.c2 {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
  cursor: pointer;
  z-index: 1100;
}

@media (max-width:calc(28.375rem + 2rem)) {
  .c1 {
    width: calc(100% - 2rem);
  }
}

<div>
  <div
    class="sc-AxhCb c0 dimming-fade-animation-appear dimming-fade-animation-appear-active"
    data-testid="dimming-background"
  />
  <div
    class="sc-AxjAm c1 centered-card-animation-appear centered-card-animation-appear-active"
    width="28.375rem"
  >
    <svg
      class="c2"
      color="#001E3E"
      data-testid="close-icon"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        fill="currentColor"
        fill-rule="nonzero"
      />
    </svg>
    Content
  </div>
</div>
`;

exports[`Offcanvas renders the top side variation 1`] = `
.c1 {
  overflow: auto;
  box-sizing: border-box;
  background-color: #FFFFFF;
  border-radius: 0.125rem;
  padding: 2.5rem;
  height: 28.375rem;
  box-shadow: 0 0 0.75rem 0.25rem rgba(0,0,0,0.12);
  position: fixed;
  left: 0;
  right: unset;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
  z-index: 1100;
  width: 100%;
  will-change: transform,opacity;
  -webkit-transition: -webkit-transform 150ms ease-out,opacity 150ms ease-out;
  -webkit-transition: transform 150ms ease-out,opacity 150ms ease-out;
  transition: transform 150ms ease-out,opacity 150ms ease-out;
  left: 0;
  right: 0;
  top: 0;
  bottom: unset;
}

.c1.centered-card-animation-appear,
.c1.centered-card-animation-enter {
  opacity: 0;
  -webkit-transform: translate(0%,calc(-50% - 1rem));
  -ms-transform: translate(0%,calc(-50% - 1rem));
  transform: translate(0%,calc(-50% - 1rem));
}

.c1.centered-card-animation-appear-active,
.c1.centered-card-animation-enter-active {
  opacity: 1;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
}

.c1.centered-card-animation-exit {
  opacity: 1;
  -webkit-transform: translate(0%,0%);
  -ms-transform: translate(0%,0%);
  transform: translate(0%,0%);
}

.c1.centered-card-animation-exit-done,
.c1.centered-card-animation-exit-active {
  opacity: 0;
  -webkit-transform: translate(-0%,-50%);
  -ms-transform: translate(-0%,-50%);
  transform: translate(-0%,-50%);
}

.c0 {
  background-color: rgba(0,15,31,0.6);
  height: 100%;
  left: 0;
  opacity: 0.6;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1000;
  will-change: opacity;
  -webkit-transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
  transition: opacity 200ms cubic-bezier(0.215,0.61,0.355,1);
}

.c0.dimming-fade-animation-appear,
.c0.dimming-fade-animation-enter {
  opacity: 0;
}

.c0.dimming-fade-animation-appear-active,
.c0.dimming-fade-animation-enter-active {
  opacity: 0.6;
}

.c0.dimming-fade-animation-exit {
  opacity: 0.6;
}

.c0.dimming-fade-animation-exit-done,
.c0.dimming-fade-animation-exit-active {
  opacity: 0;
}

.c2 {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
  cursor: pointer;
  z-index: 1100;
}

@media (max-width:calc( + 2rem)) {
  .c1 {
    width: calc(100% - 2rem);
  }
}

<div>
  <div
    class="sc-AxhCb c0 dimming-fade-animation-appear dimming-fade-animation-appear-active"
    data-testid="dimming-background"
  />
  <div
    class="sc-AxjAm c1 centered-card-animation-appear centered-card-animation-appear-active"
    height="28.375rem"
  >
    <svg
      class="c2"
      color="#001E3E"
      data-testid="close-icon"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        fill="currentColor"
        fill-rule="nonzero"
      />
    </svg>
    Content
  </div>
</div>
`;
